svm/src/account_loader.rs

```diff
@@ -95,6 +96,264 @@ pub struct FeesOnlyTransaction {
+}
+impl<'a, CB: TransactionProcessingCallback> AccountLoader<'a, CB> {
+    pub fn new(
+        callbacks: &'a CB,
+        account_overrides: Option<&'a AccountOverrides>,
+        program_cache: ProgramCacheForTxBatch,
+    ) -> AccountLoader<'a, CB> {
+        Self {
+            program_cache,
+            account_overrides,
+            account_cache: HashMap::new(),
+            callbacks,
+        }
+    }
+
+    pub fn load_account(
+        &mut self,
+        account_key: &Pubkey,
+        usage_pattern: AccountUsagePattern,
+    ) -> Option<LoadedTransactionAccount> {
+        let is_writable = usage_pattern.is_writable();
+        let is_invisible = usage_pattern.is_invisible();
+
+        if let Some(cache_item) = self.account_cache.get_mut(account_key) {
+            if !cache_item.inspected_as_writable {
+                // Inspecting a read-only account is harmless. Inspecting a writable
+                // account must be done before rent collection, and must never be done
+                // again in-batch, because inspection is intended to preserve the
+                // state prior to any writes.
+                self.callbacks.inspect_account(
+                    account_key,
+                    AccountState::Alive(&cache_item.account),
+                    is_writable,
+                );
+                cache_item.inspected_as_writable = is_writable;
+            }
+
+            // We must compare the full account, not just check if lamports are 0.
+            // Otherwise we might erroneously hide rent-delinquent read-only accounts.
+            if cache_item.account == AccountSharedData::default() {
+                None
+            } else if let Some(program) = is_invisible
+                .then_some(())
+                .and_then(|_| self.program_cache.find(account_key))
+            {
+                // NOTE if this account could be found in the program cache, we must
+                // substitute its size here, to perserve an oddity with transaction size
+                // calculations. However, we CANNOT skip the accounts cache check,
+                // in case the account was non-executable and modified in-batch.
+                // When we properly check if program cache entries are executable,
+                // we can go to the program cache first and remove this branch, because
+                // executable accounts are immutable.
+                Some(LoadedTransactionAccount {
+                    loaded_size: program.account_size,
+                    account: account_shared_data_from_program(&program),
+                    rent_collected: 0,
+                })
+            } else {
+                Some(LoadedTransactionAccount {
+                    loaded_size: cache_item.account.data().len(),
+                    account: cache_item.account.clone(),
+                    rent_collected: 0,
+                })
+            }
+        } else if let Some(account_override) = self
```

Comment:
it seems like account overrides should go before checking the cache. otherwise only the first transaction would actually override it, right?

Reply:
who uses these anyway? is there an actual use case for these or are they just over-complicating stuff?

Reply:
nope, since the accounts db branch comes before the overrides branch, and the overrides branch puts it in cache. so the first transaction would get it from overrides and put the overridden version in cache, the second transaction would get the overriden version from cache, including any changes the first transaction might have made to it

Reply:
i knew offhand it was used for transaction simulation, but i looked into it more since you asked and the only thing that appears to ever be overridden is the slot history sysvar. since sysvars cant be modified, the flow to propagate updated overridden accounts shouldnt matter, but i think we should leave it like this for this pr, and explore a better way for the runtime to pass this in in the future (as long as i havent missed any other uses of `AccountOverrides`)

Reply:
For sure, just always looking for code to simplify or split out if it is not necessary for validator operation

Reply:
Just for awareness, the jito fork uses account overrides as an intra-bundle accounts cache. cc @buffalu

Reply:
I think Jito can use this implementation instead of using account overrides though. Let's them reduce their code fork changes, even if until simd83 is activated they still need to record each tx separately.